firestore: DataTo() Inconsistent if Field Names Differ by First Character

**Description**

I have a static Firestore Document with two fields differing by the first letter: `tags` and `Tags`.  On some reads, the array is added to the slice but other times the slice is empty. It's easy to reproduce this issue by rerunning the `DataTo()` method a couple times.

**Client**

e.g. Cloud Run

**Environment**

e.g. Scratch Docker on Cloud Run

**Go Environment**

$ go version go1.17 darwin/amd64

**Code**

e.g.
```go
package main

import (
	"context"
	"log"
	"os"

	"cloud.google.com/go/firestore"
)

type test struct {
	Nothing string
	Tags    []string
}

func main() {
	ctx := context.Background()

	client, err := firestore.NewClient(ctx, os.Getenv("GOOGLE_CLOUD_PROJECT"))
	if err != nil {
		log.Fatal(err)
	}

	docRef := client.Doc("Collection/test")

	if _, err := docRef.Set(ctx, map[string]interface{}{"Tags": nil, "tags": []string{"value1", "value2"}}, firestore.MergeAll); err != nil {
		log.Fatal(err)
	}

	for i := 0; i < 20; i++ {
		docSnap, err := docRef.Get(ctx)
		if err != nil {
			log.Fatal(err)
		}

		value := &test{}

		if err := docSnap.DataTo(value); err != nil {
			log.Fatal(err)
		}

		log.Printf("%+v", value)
	}

}
```

**Expected behavior**

"MyArray" slice always includes each element in the Firestore array.
```sh
$ go run ./test/main.go 
2021/09/02 15:31:02 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:02 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:02 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:02 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:02 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:02 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:04 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:04 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:03 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:03 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:03 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:04 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:03 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:03 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:03 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:04 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:03 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:04 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:04 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:04 &{Nothing: Tags:[value1 value2]}
```

**Actual behavior**

"MayArray" is sometimes empty on a query.
```sh
$ go run ./test/main.go 
2021/09/02 15:31:02 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:02 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:02 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:02 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:02 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:02 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:02 &{Nothing: Tags:[]}
2021/09/02 15:31:02 &{Nothing: Tags:[]}
2021/09/02 15:31:03 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:03 &{Nothing: Tags:[]}
2021/09/02 15:31:03 &{Nothing: Tags:[]}
2021/09/02 15:31:03 &{Nothing: Tags:[]}
2021/09/02 15:31:03 &{Nothing: Tags:[]}
2021/09/02 15:31:03 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:03 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:03 &{Nothing: Tags:[]}
2021/09/02 15:31:03 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:04 &{Nothing: Tags:[]}
2021/09/02 15:31:04 &{Nothing: Tags:[value1 value2]}
2021/09/02 15:31:04 &{Nothing: Tags:[value1 value2]}
```

Confirmed this behavior
```
func TestIntegration_DataTo_4722(t *testing.T) {
	ctx := context.Background()
	coll := integrationColl(t)

	type test struct {
		Nothing string
		Tags    []string
	}

	docRef := coll.NewDoc()
	if _, err := docRef.Set(ctx, map[string]interface{}{"Tags": nil, "tags": []string{"value1", "value2"}}, MergeAll); err != nil {
		log.Fatal(err)
	}

	for i := 0; i < 20; i++ {
		docSnap, err := docRef.Get(ctx)
		if err != nil {
			log.Fatal(err)
		}

		value := &test{}

		if err := docSnap.DataTo(value); err != nil {
			log.Fatal(err)
		}

		if value.Tags != nil {
			t.Errorf("Unexpected content on 'Tags': iteration: %d, got: %s want: nil", i, value.Tags)
		}
	}
}

```

```

--- FAIL: TestIntegration_DataTo_4722 (3.49s)
    /Users/crwilcox/workspace/google-cloud-go-3/firestore/integration_test.go:1742: Unexpected content on 'Tags': iteration: 10, got: [value1 value2] want: nil
    /Users/crwilcox/workspace/google-cloud-go-3/firestore/integration_test.go:1742: Unexpected content on 'Tags': iteration: 11, got: [value1 value2] want: nil
    /Users/crwilcox/workspace/google-cloud-go-3/firestore/integration_test.go:1742: Unexpected content on 'Tags': iteration: 12, got: [value1 value2] want: nil
    /Users/crwilcox/workspace/google-cloud-go-3/firestore/integration_test.go:1742: Unexpected content on 'Tags': iteration: 13, got: [value1 value2] want: nil
    /Users/crwilcox/workspace/google-cloud-go-3/firestore/integration_test.go:1742: Unexpected content on 'Tags': iteration: 15, got: [value1 value2] want: nil
    /Users/crwilcox/workspace/google-cloud-go-3/firestore/integration_test.go:1742: Unexpected content on 'Tags': iteration: 16, got: [value1 value2] want: nil
    /Users/crwilcox/workspace/google-cloud-go-3/firestore/integration_test.go:1742: Unexpected content on 'Tags': iteration: 19, got: [value1 value2] want: nil
FAIL
FAIL	cloud.google.com/go/firestore	4.214s
FAIL
```

Worth sharing also, DataTo ran on the same data can vary... So this doesn't seem to be an issue with what is retrieved from the service. 


```
func TestIntegration_DataTo_4722(t *testing.T) {
	ctx := context.Background()
	coll := integrationColl(t)

	type test struct {
		Nothing string
		Tags    []string
	}

	docRef := coll.NewDoc()
	if _, err := docRef.Set(ctx, map[string]interface{}{"Tags": nil, "tags": []string{"value1", "value2"}}, MergeAll); err != nil {
		log.Fatal(err)
	}

	for i := 0; i < 20; i++ {
		docSnap, err := docRef.Get(ctx)
		if err != nil {
			log.Fatal(err)
		}
		value := &test{}
		if err := docSnap.DataTo(value); err != nil {
			log.Fatal(err)
		}

		value1 := &test{}
		if err := docSnap.DataTo(value1); err != nil {
			log.Fatal(err)
		}

		if cmp.Diff(value, value1) != "" {
			t.Errorf("DataTo should be identical: first: %+v second: %+v", value, value1)
		}
	}
}
```
```
--- FAIL: TestIntegration_DataTo_4722 (3.52s)
    /Users/crwilcox/workspace/google-cloud-go-3/firestore/integration_test.go:1742: DataTo should be identical: first: &{Nothing: Tags:[value1 value2]} second: &{Nothing: Tags:[]}
    /Users/crwilcox/workspace/google-cloud-go-3/firestore/integration_test.go:1742: DataTo should be identical: first: &{Nothing: Tags:[value1 value2]} second: &{Nothing: Tags:[]}
    /Users/crwilcox/workspace/google-cloud-go-3/firestore/integration_test.go:1742: DataTo should be identical: first: &{Nothing: Tags:[value1 value2]} second: &{Nothing: Tags:[]}
    /Users/crwilcox/workspace/google-cloud-go-3/firestore/integration_test.go:1742: DataTo should be identical: first: &{Nothing: Tags:[value1 value2]} second: &{Nothing: Tags:[]}
    /Users/crwilcox/workspace/google-cloud-go-3/firestore/integration_test.go:1742: DataTo should be identical: first: &{Nothing: Tags:[]} second: &{Nothing: Tags:[value1 value2]}
FAIL
```

--- a/firestore/document_test.go
+++ b/firestore/document_test.go
@@ -160,6 +160,38 @@ func TestDataAt(t *testing.T) {
 	}
 }
 
+func TestDataTo(t *testing.T) {
+	doc := &DocumentSnapshot{
+		proto: &pb.Document{
+			Fields: map[string]*pb.Value{
+				"tags": arrayval(strval("value1"), strval("value2")),
+				"Tags": arrayval(strval("value3"), strval("value4")),
+				"TaGs": {ValueType: &pb.Value_NullValue{}},
+			},
+		},
+	}
+
+	type test struct {
+		Nothing string
+		Tags    []string
+	}
+
+	// DataTo displayed indeterminate results run to run.
+	// https://github.com/googleapis/google-cloud-go/issues/4722
+	want := &test{Tags: []string{"value3", "value4"}}
+
+	for i := 0; i < 20; i++ {
+		got := &test{}
+		if err := doc.DataTo(got); err != nil {
+			t.Fatal(err)
+		}
+
+		if !testEqual(got, want) {
+			t.Fatalf("got %#v\nwant %#v", got, want)
+		}
+	}
+}
+
 func TestDataAtPath(t *testing.T) {
 	for _, test := range []struct {
 		fieldPath FieldPath

--- a/firestore/from_value.go
+++ b/firestore/from_value.go
@@ -20,6 +20,7 @@ import (
 	"reflect"
 	"strings"
 
+	"cloud.google.com/go/internal/fields"
 	"github.com/golang/protobuf/ptypes"
 	pb "google.golang.org/genproto/googleapis/firestore/v1"
 )
@@ -306,15 +307,19 @@ func createMapFromValueMap(pm map[string]*pb.Value, c *Client) (map[string]inter
 // populateStruct sets the fields of vs, which must be a struct, from
 // the matching elements of pm.
 func populateStruct(vs reflect.Value, pm map[string]*pb.Value, c *Client) error {
-	fields, err := fieldCache.Fields(vs.Type())
+	fs, err := fieldCache.Fields(vs.Type())
 	if err != nil {
 		return err
 	}
 
+	type match struct {
+		vproto *pb.Value
+		f      *fields.Field
+	}
 	// Find best field matches
-	matched := make(map[string]*pb.Value)
+	matched := make(map[string]match)
 	for k, vproto := range pm {
-		f := fields.Match(k)
+		f := fs.Match(k)
 		if f == nil {
 			continue
 		}
@@ -322,19 +327,18 @@ func populateStruct(vs reflect.Value, pm map[string]*pb.Value, c *Client) error 
 			// If multiple case insensitive fields match, the exact match
 			// should win.
 			if f.Name == k {
-				matched[k] = vproto
+				matched[k] = match{vproto: vproto, f: f}
 			}
 		} else {
-			matched[f.Name] = vproto
+			matched[f.Name] = match{vproto: vproto, f: f}
 		}
 	}
 
 	// Reflect values
-	for k, vproto := range matched {
-		f := fields.Match(k)
-		if f == nil {
-			continue
-		}
+	for _, v := range matched {
+		f := v.f
+		vproto := v.vproto
+
 		if err := setReflectFromProtoValue(vs.FieldByIndex(f.Index), vproto, c); err != nil {
 			return fmt.Errorf("%s.%s: %v", vs.Type(), f.Name, err)
 		}

--- a/firestore/from_value.go
+++ b/firestore/from_value.go
@@ -310,7 +310,27 @@ func populateStruct(vs reflect.Value, pm map[string]*pb.Value, c *Client) error 
 	if err != nil {
 		return err
 	}
+
+	// Find best field matches
+	matched := make(map[string]*pb.Value)
 	for k, vproto := range pm {
+		f := fields.Match(k)
+		if f == nil {
+			continue
+		}
+		if _, ok := matched[f.Name]; ok {
+			// If multiple case insensitive fields match, the exact match
+			// should win.
+			if f.Name == k {
+				matched[k] = vproto
+			}
+		} else {
+			matched[f.Name] = vproto
+		}
+	}
+
+	// Reflect values
+	for k, vproto := range matched {
 		f := fields.Match(k)
 		if f == nil {
 			continue

